### PR TITLE
feat(deepseek-v41): add context parallelism and packed SFT

### PR DESCRIPTION
**Pull request by HuiyingLi**
# What does this PR do ?

Add context-parallel packed SFT for the DeepSeek V4.1-Flash text model, preserving encoder compressed-KV reuse and the encoder-to-decoder handoff. The TileLang path also removes unused dense attention masks and fuses indexer scoring to reduce long-context temporary memory.

This branch merges main through ba51161a, including the fused Triton gradient-norm implementation from #3685. Both DS41 CP recipes explicitly select the shared backend:

```yaml
clip_grad_norm:
  max_norm: 1.0
  backend: triton
```

# Changelog

- **Context parallelism:** contiguous query shards, global token positions for RoPE/causality, autograd-aware KV gathering, and source-state propagation through encoder reuse/reindex layers and the decoder handoff.
- **One packed input path:** row-shaped tokens with `seq_lens` / `seq_lens_padded`; compression-aligned physical spans counted inside the pack budget. Preserve independently shifted recipe labels and supervised-token denominators.
- **Document isolation:** reset RoPE positions per document; carry document IDs through window/compressed attention, candidate selection and Engram history. Cover documents spanning CP ranks and ranks containing only padding.
- **Indexer memory:** TileLang scoring fuses dot product, ReLU, head weighting and reduction while preserving BF16 rounding and packed/CP visibility. Avoid materializing per-head score intermediates. Construct dense attention masks only for paths that consume them.
- **Gradient norm:** use main's shared Triton FP64 reducer via clip_grad_norm.backend: triton. The shared norm implementation and LLM training entry point are identical to main; no branch-specific norm implementation remains. backend: te selects the upstream TE alternative.
- **Recipes and tests:** add unpacked Tulu3 4K CP and packed Tulu3 CP8 32K recipes, plus semantic packing, CP/autograd, TileLang and grad-norm coverage.

## Validation and limitations

Results below are from separate pinned runs. The historical CP1/CP2, CP8 and TE OFF/ON runs predate this main merge and do not establish full-model training parity for the Triton configuration. This update is validated separately by owner-shape memory/numeric measurements, recipe forwarding, two-GPU norm/clipping/FSDP checks and CP/packed regression tests. Original experiment artifacts remain on the cluster.

### Triton gradient norm: actual DS41 owner-shard shapes

Merged-branch validation: **67 recipe/norm unit tests, 25 CUDA/DTensor/FSDP tests and 15 CP/packed tests passed**. The distributed norm cases exercise both TE and Triton, including empty/uneven shards, Replicate/Partial placements, mixed local layouts/counts, PP scalar reduction, and a real FSDP backward/clipping/optimizer update. The upstream norm suite separately passed 44 tests. Pre-commit and YAML checks passed.

On one GB200 in the reused interactive allocation, compare synthetic BF16 gradients at the two actual EP64 owner-shard shapes (6,000,097 x 256 and 6,000,261 x 256). This measures **local norm plus in-place clipping**, after two warmups, median of 20 synchronized repetitions.

| Backend | Extra PyTorch allocator peak | Median norm + clipping | Relative norm error vs analytic reference |
| --- | ---: | ---: | ---: |
| Historical PyTorch reduction | 17.168 GiB | 40.42 ms | 0 |
| Upstream TE | 4 KiB | 12.41 ms | 4.01e-8 |
| Upstream Triton | 735 KiB | 10.74 ms | 0 |

All repeated norms were bitwise stable within each backend, and clipped gradient samples matched the reference exactly. Both fused backends directly consume contiguous local gradient storage; Triton widens values inside the kernel and allocates chunk partials instead of full-size gradient temporaries. These are isolated workspace measurements, not full-training peak memory or step time. No new multinode training run was submitted for this update.

### Semantic tests and fused-indexer replacement

- Historical integration suite: **92 tests passed**; Ruff and diff checks passed.
- Packed versus independent documents: logits, intermediate layers, supervised loss, gradients and an optimizer update; cross-document isolation; complete 40-layer source/reuse/reindex schedule in reduced-width FP32 tests.
- Real CP2 tests with activation checkpointing ON/OFF, odd document lengths, rank-boundary crossings and padding-only ranks. Separate GPU functional cases cover the six attention roles, reused state, candidates and reindexing.
- Real pretrained first four layers, full width and 4K input: **fused versus original indexer logits are bitwise identical** on all four ranks.
- Strict first-four-layer official-reference gates remain failed identically before/after the replacement: mean KL 0.0062183 (limit 0.001), p95 KL 0.0357104 (limit 0.002), top-1 agreement 94.5068% (required 98%). This is not official-reference parity.
- A separate small six-layer BF16 CP1/CP2 diagnostic exceeded its 1% logit relative-RMSE gate (observed 1.63%); the failure remains recorded.
- Isolated CP8-shaped 32K scoring: median 14.875 ms / 24.00 GiB extra allocation with original scoring versus 5.759 ms / 0.25 GiB with fused scoring. Top-512 indices and visibility matched exactly; five finite scores differed within the one-BF16-ULP gate. These measurements exclude projections, communication, top-k and full training.

### Full pretrained packed Tulu3 4K: CP1 versus CP2

Historical jobs 7101485 / 7101490 each completed **100 updates** with the full 40-layer model, EP64/world64, GBS64/LBS1 and per-block AC. Initial parameter samples, all 100 input audits, LR and supervised-token counts matched. Both native W&B histories have 100 rows, all original logger values exact.

| Gate | Observed | Threshold / result |
| --- | ---: | --- |
| Maximum training-loss difference | 0.001736820 | <= 0.01: pass |
| Mean training-loss difference | 0.000554583 | <= 0.002: pass |
| Maximum validation-loss difference | 0.000443995 | <= 0.01: pass |
| Separate maximum relative gradient-norm diagnostic | 27.1212% | <= 5%: **fail** |

Primary training gates passed; exact BF16 gradient parity is not established.

W&B: [packed CP1](https://wandb.ai/Nemo-automodel/huiyingl_workspace/runs/ds41packedcp17101485), [packed CP2](https://wandb.ai/Nemo-automodel/huiyingl_workspace/runs/ds41packedcp27101490).

image

### Full pretrained packed CP8 32K

The original indexer attempt 7112881 completed 1 update before an indexer-scoring OOM. The fused-indexer rerun 7118100 completed **70/100 updates without observed OOM**, then hit its single one-hour allocation limit. Final completed training loss 0.570156; validation loss 0.586528 at index 49. 

[Native W&B](https://wandb.ai/Nemo-automodel/huiyingl_workspace/runs/ds41packed32kcp820260913fused) contains 69 exact rows (indices 0–68); completed local index 69 is missing online. The native-to-W&B completeness gate remains failed. No resumable checkpoint was produced and no continuation was run. This run predates the TE opt-in integration.
image

### Latest full pretrained packed 4K CP2: TE OFF versus ON

Same frozen `30f41b61` source, global 4096 / local 2048, CP2/EP64/world64, 16 nodes x 4 GPUs, GBS64/LBS1, per-microbatch FSDP synchronization. Only `clip_grad_norm.use_te` differs in the training configuration.

| Variant | Job | Completed updates | Outcome | Native W&B |
| --- | --- | ---: | --- | --- |
| OFF | 7128311 | 2/100 | Third attempted update OOM in PyTorch grad norm | [2 exact rows, failed](https://wandb.ai/Nemo-automodel/automodel-ds41-gradnorm-packed4k-20260913/runs/ds41te4k07128311) |
| ON | 7128312 | 100/100 | COMPLETED, exit 0, 19m17s within 30min | [100 exact rows, finished](https://wandb.ai/Nemo-automodel/automodel-ds41-gradnorm-packed4k-20260913/runs/ds41te4k17128312) |

ON final training / validation loss: **0.571274 / 0.546551**. The common two-update prefix has maximum loss difference 0.000245035 and maximum relative norm difference 0.0303389%; both pass the prefix thresholds. **Full 100-step OFF/ON parity is not established** because OFF ended early; completion, step-count, full-input-audit and validation comparison gates remain failed.

OFF's norm reduction failed while requesting 11.45 GiB with 10.81 GiB free on a 184.31 GiB GPU. Main-rank native `mem` is a PyTorch allocator peak over the logging interval, not total device usage or isolated norm workspace; its reported maximum is 132.07866 GiB OFF / 132.08661 GiB ON. The failed third OFF update has no completed log row. This is an OOM-versus-completion observation, not a measured full-training peak-memory delta.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Followed contributor workflow; all branch commits include DCO sign-off.
- [x] Added numerical, packed-sequence and distributed gradient tests.
- [x] Added user-facing recipe documentation; local experiment outcomes and W&B links are summarized in this PR body.
- [ ] Resolve or explicitly disposition the retained numerical limitations before declaring broader parity.

# Additional Information

Shared gradient-norm implementation: merged #3685 (Triton and TE backends). Historical TE comparison: #3878. No DSV4 model implementation files are changed. CP currently supports text `input_ids`; image packing and preflattened `cu_seqlens` input are outside this path. The packed 4K, first-four-layer, CP8 32K and latest TE comparisons have distinct validation scopes.

### CODE REVIEW

File: docs/model-coverage/llm/deepseek-ai/deepseek-v41-flash.mdx
Comment (jgerh) [inline docs/model-coverage/llm/deepseek-ai/deepseek-v41-flash.mdx:3]:
```suggestion
description: "DeepSeek-V4.1 Flash backbone fine-tuning with expert parallelism and sharded Engram memory."
```

Comment (jgerh) [inline docs/model-coverage/llm/deepseek-ai/deepseek-v41-flash.mdx:17]:
```suggestion
| Architecture | `DeepseekV41ForCausalLM` |
```

File: docs/model-coverage/llm/deepseek-ai/deepseek-v41-flash.mdx [hunk 1]
```diff
@@ -67,6 +67,29 @@ See the [Slurm launcher guide](/job-launchers/slurm-cluster) and
...
+
+Queries and hidden activations remain local. Window KV, newly compressed KV,
+and index keys are gathered across the CP group; reused compressed KV retains
+its gradient history across layers. Engram gathers integer token history for
+N-gram hashing, then looks up only the local tokens through the existing row
+owners. CP does not gather full Engram tables. TP, PP, packed documents, and
+image inputs remain outside this CP path.
+
+The [Tulu3 CP recipe](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/deepseek_v41/deepseek_v41_flash_tulu3_cp.yaml)
+uses 4,096-token sequences, TE RMSNorm, EP64 and 64 Engram owners. At global
```
Comment (jgerh) [inline docs/model-coverage/llm/deepseek-ai/deepseek-v41-flash.mdx:87]:
```suggestion
uses 4,096-token sequences, TE RMSNorm, EP64, and 64 Engram owners. At global
```

File: docs/model-coverage/llm/deepseek-ai/deepseek-v41-flash.mdx
Comment (jgerh) [inline docs/model-coverage/llm/deepseek-ai/deepseek-v41-flash.mdx:143]:
```suggestion
Official TP1 was compared with native EP4 and 4 Engram owners using
```

Comment (jgerh) [inline docs/model-coverage/llm/deepseek-ai/deepseek-v41-flash.mdx:169]:
```suggestion
combine. It used 4 GB200 GPUs, the released first four layers and full
```

**Review by jgerh — COMMENTED** [review]
Completed tech pubs review of docs/model-coverage/llm/deepseek-ai/deepseek-v41-flash.mdx and provided a few minor copyedits.

**Review by akoumpa — APPROVED** [review]
